src/arcsf/config/experiment.py

```diff
@@ -0,0 +1,304 @@
+        obj: Object to ensure is a list.
+
+    Returns:
+        Either the original object if a list or a list containing the object.
+    """
+    if isinstance(obj, list):
+        return obj
+    return [obj]
+
+
+def _make_config_name(
+    top_config_name: str,
+    n: int,
+) -> str:
+    return f"{top_config_name}/experiment_{n}"
+
+
+def _generate_combos_from_dict(
+    combo_dict: dict,
+    full: bool,
+) -> dict:
+    # Convert combo dict into dict of combos
+    sweep_dict_keys, sweep_dict_vals = zip(*combo_dict.items())
+    combinations = [
+        dict(zip(sweep_dict_keys, v)) for v in product(*list(sweep_dict_vals))
+    ]
+
+    # Initialise train_type
+    train_type = "retain"
+    if full:
+        train_type = "full"
+
+    # Add train_type to each combination
+    for n, _ in enumerate(combinations):
+        combinations[n]["train_type"] = train_type
+
+    # Return
+    return combinations
+
+
+def generate_combos_from_dict(
+    combo_dict: dict,
+    full: bool,
+    wandb_kwargs: dict,
+) -> dict:
+    # Generate combinations
+    combinations = _generate_combos_from_dict(combo_dict, full)
+
+    # Process model kwargs
+    for n, _ in enumerate(combinations):
+        combinations[n]["hyperparameter_config"] = combinations[n]["model_config"][1]
+        combinations[n]["model_config"] = combinations[n]["model_config"][0]
+        combinations[n] = {**combinations[n], **wandb_kwargs}
+
+    # Return
+    return combinations
+
+
+def generate_experiment_configs(
+    top_config_name: str,
+) -> None:
+
+    # Read in yaml file
+    with open(
+        os.path.join(arcsf.constants.EXPERIMENT_CONFIG_DIR, f"{top_config_name}.yaml")
+    ) as f:
+        top_config = yaml.safe_load(f)
+
+    # Loop over, construct combo dict
+    combo_dict = {}
+    for key, value in top_config["combinations"].items():
+        combo_dict[key] = _listify(value)
+
+    # Construct same dict with full dataset
+    combo_dict_full = copy(combo_dict)
+    combo_dict_full["data_config"] = [top_config["full_data_config"]]
+
+    # Get wandb kwargs
+    wandb_kwargs = {}
+    if "wandb_kwargs" in top_config.keys():
+        wandb_kwargs = {**top_config["wandb_kwargs"]}
+
+    # Get combinations
+    full_combinations = generate_combos_from_dict(combo_dict_full, True, wandb_kwargs)
+    retain_combinations = generate_combos_from_dict(combo_dict, False, wandb_kwargs)
+
+    # Add full config path to
+    for n, retain_combo in enumerate(retain_combinations):
+        for m, full_combo in enumerate(full_combinations):
+            if (
+                full_combo["model_config"] == retain_combo["model_config"]
+                and full_combo["seed"] == retain_combo["seed"]
+                and full_combo["hyperparameter_config"]
+                == retain_combo["hyperparameter_config"]
+            ):
+                retain_combinations[n]["full_model_name"] = _make_config_name(
+                    top_config_name, m
+                )
+
```

Comment:
Not sure how much of an improvement it is in the end but was wondering whether some of this could be refactored along the lines of the below to avoid needing the big `if` at the end to recover the full model id (not tested + some of the other functions would need to be changed slightly):

```python
data_configs = top_config["combination"].pop("data_config")
combinations = generate_combos_from_dict(top_config["combination"], wandb_args):
retain_configs = []
full_configs = []
for full_idx, c in enumerate(combinations):
    # Model trained on whole dataset 
    fc = deepcopy(c)
    fc["data_config"] = top_config["full_data_config"]
    fc["train_type"] = "full"
    full_configs.append(fc)

    # Models trained on different retain splits
    for dc in data_configs:
        rc = deepcopy(c)
        rc["data_config"] = dc
        rc["full_model_name"] = _make_config_name(top_config_name, full_idx)
        rc["train_type"] = "retain"
        retain_configs.append(rc)
```

Ok to leave as is if this is missing something/not an improvement.

Also, things will get messier when adding forget models, I guess that would be an extra nested for loop...